cli/src/commands/diff.rs

```diff
@@ -52,38 +59,60 @@ pub(crate) struct DiffArgs {
     format: DiffFormatArgs,
 }
 
+fn resolve_revision(
+    workspace_command: &WorkspaceCommandHelper,
+    r: &Option<RevisionArg>,
+) -> Result<Commit, CommandError> {
+    workspace_command.resolve_single_rev(r.as_ref().unwrap_or(&RevisionArg::AT))
+}
+
 #[instrument(skip_all)]
 pub(crate) fn cmd_diff(
     ui: &mut Ui,
     command: &CommandHelper,
     args: &DiffArgs,
 ) -> Result<(), CommandError> {
     let workspace_command = command.workspace_helper(ui)?;
-    let from_tree;
+    let diff_renderer = workspace_command.diff_renderer_for(&args.format)?;
+    let fileset_expression = workspace_command.parse_file_patterns(&args.paths)?;
+    let matcher = fileset_expression.to_matcher();
+
     let to_tree;
-    if args.from.is_some() || args.to.is_some() {
-        let from =
-            workspace_command.resolve_single_rev(args.from.as_ref().unwrap_or(&RevisionArg::AT))?;
-        from_tree = from.tree()?;
-        let to =
-            workspace_command.resolve_single_rev(args.to.as_ref().unwrap_or(&RevisionArg::AT))?;
+    let from_tree;
+    let mut copy_records = CopyRecordMap::default();
+
+    if args.to.is_some() || args.from.is_some() {
+        let to = resolve_revision(&workspace_command, &args.to)?;
+        let from = resolve_revision(&workspace_command, &args.from)?;
+        copy_records.add_records(workspace_command.repo().store().get_copy_records(
+            &[],
+            from.id(),
+            to.id(),
+        )?)?;
         to_tree = to.tree()?;
+        from_tree = from.tree()?;
     } else {
-        let commit = workspace_command
-            .resolve_single_rev(args.revision.as_ref().unwrap_or(&RevisionArg::AT))?;
-        from_tree = commit.parent_tree(workspace_command.repo().as_ref())?;
-        to_tree = commit.tree()?
+        let to = resolve_revision(&workspace_command, &args.revision)?;
+        let parents: Vec<_> = to.parents().try_collect()?;
+        for p in &parents {
+            copy_records.add_records(workspace_command.repo().store().get_copy_records(
+                &[],
+                p.id(),
+                to.id(),
+            )?)?;
+        }
```

Comment:
Out of curiosity, how copy records are stored in a merge commit? This API suggests that the records are identified by `(parent_id, self_id)` even if `self.parents()[i] == parent`.

Suppose a file is renamed at one side of the merge,
```
D
|\
C |  rename "foo" -> "bar", modify "bar"
| |
| B  modify "foo"
: :
|/
A
```
when creating the merge commit `D` by `jj new C B`, auto-merge function would gather copy records of `A::C` and `A::B`, then merge file `D["bar"] = C["bar"] - A["foo"] + B["foo"]`. I'm not sure what would be recorded at `D`. If the auto-merged tree was built considering renames, the commit `D` would have no copy records (unless user made another rename in addition)?

If the backend doesn't have persistent copy records, I guess both `trace_copy(C to D)` and `trace_copy(B to D)` would report `"foo" -> "bar"`.

Reply:
this is currently left to `CopyRecordMap` to handle conflicts (which discards both copy traces).  Unfortunately, this is almost purely theoretical as git doesn't keep copy traces but infers them dynamically based on content.  I think that properly solving conflicting copies is going to have to be a follow on thing handled by Google folks, since their backend is the only one that can manifest them.

Reply:
> I'm not sure what would be recorded at `D`.

Good question! We just don't know yet (or at least I don't know yet). I'm going to chat with @ilyagr and @torquestomp about this next week. We'll need to be able to reproduce the parent tree somehow. Maybe that means we'll need to store the copy information from `A::C` and `A::B`. We will also want to be able to represent conflicts in the copy information state itself (e.g. rename `foo -> bar` on one side and `foo -> baz` on the other side). And we'll want to be able to back out such a commit. Lots of unanswered questions :(

Reply:
I wonder if this was the reason that `roots` in `get_copy_records()` was a list. For a linear history `A - B - C`, `get_copy_records(A, C)` would accumualte copy records of `B` and `C`. `A`'s records are not included because `A` is the base tree. Given that, for a merge `{D, E} - F`, both `get_copy_records(D, F)` and `get_copy_records(E, F)` would return the copy records of `F`, and should be a single query `get_copy_records([D, E], F)`?

Reply:
I am going to resolve this conversation as out of scope for this PR.  Hopefully folks can figure out the details next week and build on this though!

Reply:
Oh, I thought this would matter whether the first commit of this PR makes sense or not.

Maybe https://github.com/martinvonz/jj/pull/4235 is a better place to discuss this kind of things, but appears that #4235 is a blocker of this PR.

Reply:
I also don't think the first commit really makes sense, but there's a lot of work after it that I don't want to block. I'm also not sure how the API should look. I.e., I'm not sure that the state before the first commit was right either. I agree with you that the reason it was a list was probably for merge commits, but I'm still not sure how to interpret it. I think @torquestomp and I have been out at different times so I haven't gotten a chance to ask him about it either.

And I would have preferred for #4235 to have been part of the first commit, btw.

Reply:
Agreed. I'm not strongly against merging this PR. It can be easily reverted, and people here (including me) wouldn't be an expert of copy tracking.

> And I would have preferred for https://github.com/martinvonz/jj/pull/4235 to have been part of the first commit, btw.

Suppose the new API isn't right, I would rather leave the current design doc as is.

Reply:
Sorry for being late on this thread. I asked for the extraction of the design doc changes from this PR as I thought we'd run into such a problem. So it was the easiest way to unblock @fowles work and to land it in a reasonable state. 

> > And I would have preferred for #4235 to have been part of the first commit, btw.
> 
> Suppose the new API isn't right, I would rather leave the current design doc as is.

This is a great point.

Reply:
And it isn't entirely clear, what the actual API should be the implementation which we have now based on `Gix` or the proposed one from @torquestomp which needs a implementation downstream.